### DISCUSSION TITLE
typecheck: type <struct> has no field or method <Field on embedded struct from another package>

### DISCUSSION THREAD

**Discussion by snowzach** [golangci/golangci-lint#2287] [Q&A] [answered]
### Welcome

- [X] Yes, I'm using a binary release within 2 latest major releases. Only such installations are supported.
- [X] Yes, I've searched similar issues on GitHub and didn't find any.
- [X] Yes, I've included all information below (version, config, etc).
- [X] Yes, I've tried with the standalone linter if available. (https://golangci-lint.run/usage/linters/)

### Description of the problem

I am trying to run golangci-lint on this project: https://github.com/estuary/flow

It seems to be having issues resolving fields on an embedded struct where the embedded struct is from another package.
```
$ golangci-lint run
go/flowctl/cmd-ingester.go:23:42: cmd.Diagnostics undefined (type cmdIngester has no field or method Diagnostics) (typecheck)
        defer mbp.InitDiagnosticsAndRecover(cmd.Diagnostics)()
                                                ^
```
If I change the code to reference the embedded field directly it seems to resolve the issue...  ie
```
defer mbp.InitDiagnosticsAndRecover(cmd.FlowIngesterConfig.Diagnostics)()
```

### Version of golangci-lint

<details>

```console
$ golangci-lint --version
# Paste output here
```

</details>

### Configuration file

Default configuration

### Go environment

<details>

```console
go version go1.16.6 linux/amd64
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/data/zbrown/.cache/go-build"
GOENV="/data/zbrown/.config/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/data/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/data/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/linux_amd64"
GOVCS=""
GOVERSION="go1.16.6"
GCCGO="gccgo"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/data/git/est/junk/test2/go.mod"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build3806352503=/tmp/go-build -gno-record-gcc-switches"```

</details>

### Verbose output of running

<details>

```console
golangci-lint run -v
INFO [config_reader] Config search paths: [./ /data/git/est/flow /data/git/est /data/git /data / /data/zbrown]
INFO [lintersdb] Active 10 linters: [deadcode errcheck gosimple govet ineffassign staticcheck structcheck typecheck unused varcheck]
INFO [loader] Go packages loading at mode 575 (compiled_files|deps|files|types_sizes|exports_file|imports|name) took 1.217377319s
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 13.158632ms
INFO [linters context/goanalysis] analyzers took 51.737400678s with top 10 stages: buildir: 39.729403299s, inspect: 1.968549289s, nilness: 1.469968278s, printf: 1.439017276s, fact_deprecated: 1.428653125s, ctrlflow: 1.423880023s, fact_purity: 944.957097ms, SA5012: 674.113843ms, typedness: 602.028871ms, unused: 125.623659ms
INFO [runner/skip dirs] Skipped 1 issues from dir examples/soak-tests/set-ops by pattern (^|/)examples($|/)
INFO [runner/max_same_issues] 1/4 issues with text "cmd.Ingest undefined (type cmdIngester has no field or method Ingest)" were hidden, use --max-same-issues
INFO [runner] Issues before processing: 2087, after processing: 21
INFO [runner] Processors filtering stat (out/in): path_prefixer: 21/21, path_prettifier: 2087/2087, identifier_marker: 2086/2086, uniq_by_line: 22/2077, max_same_issues: 21/22, max_from_linter: 21/21, path_shortener: 21/21, filename_unadjuster: 2087/2087, exclude-rules: 2077/2086, max_per_file_from_linter: 22/22, severity-rules: 21/21, skip_file
s: 2087/2087, exclude: 2086/2086, nolint: 2077/2077, source_code: 21/21, cgo: 2087/2087, skip_dirs: 2086/2087, autogenerated_exclude: 2086/2086, diff: 22/22, sort_results: 21/21
INFO [runner] processing took 183.437428ms with stages: exclude-rules: 112.287926ms, identifier_marker: 59.256651ms, nolint: 8.788079ms, path_prettifier: 1.317435ms, skip_dirs: 501.68µs, autogenerated_exclude: 406.972µs, source_code: 331.017µs, cgo: 167.538µs, uniq_by_line: 152.088µs, filename_unadjuster: 143.039µs, max_same_issues: 49.67µs, pat
h_shortener: 8.994µs, max_from_linter: 8.334µs, max_per_file_from_linter: 6.325µs, diff: 2.564µs, sort_results: 2.316µs, exclude: 1.912µs, skip_files: 1.859µs, path_prefixer: 1.579µs, severity-rules: 1.45µs
INFO [runner] linters took 8.915159852s with stages: goanalysis_metalinter: 8.731484231s
go/flowctl/cmd-ingester.go:23:42: cmd.Diagnostics undefined (type cmdIngester has no field or method Diagnostics) (typecheck)
        defer mbp.InitDiagnosticsAndRecover(cmd.Diagnostics)()
                                                ^
...many more of the same

INFO File cache stats: 9 entries of total size 75.2KiB
INFO Memory: 103 samples, avg is 532.3MB, max is 847.1MB
INFO Execution took 10.156456715s
```
</details>

### Code example or link to a public repository

https://github.com/estuary/flow

**Comment by ldez** [comment]
Hello,

`typecheck` is not a real linter, it's just a way to display "compilation" errors.

```console
$ go build ./go/flowctl/
# github.com/mattn/go-sqlite3
sqlite3-binding.c: In function ‘sqlite3SelectNew’:
sqlite3-binding.c:128049:10: warning: function may return address of local variable [-Wreturn-local-addr]
128049 |   return pNew;
       |          ^~~~
sqlite3-binding.c:128009:10: note: declared here
128009 |   Select standin;
       |          ^~~~~~~
# github.com/jgraettinger/gorocksdb
/home/ldez/sources/go/pkg/mod/github.com/jgraettinger/gorocksdb@v0.0.0-20210726190246-aede64bf0e7b/array.go:4:11: fatal error: rocksdb/c.h: No such file or directory
    4 | // #include "rocksdb/c.h"
      |           ^~~~~~~~~~~~~
compilation terminated.
```

```console
$ go build ./go/sql-driver/
# github.com/mattn/go-sqlite3
sqlite3-binding.c: In function ‘sqlite3SelectNew’:
sqlite3-binding.c:128049:10: warning: function may return address of local variable [-Wreturn-local-addr]
128049 |   return pNew;
       |          ^~~~
sqlite3-binding.c:128009:10: note: declared here
128009 |   Select standin;
       |          ^~~~~~~
```

https://golangci-lint.run/welcome/faq/#why-do-you-have-typecheck-errors

**Comment by snowzach** [comment, marked as answer]
Ahhh Brilliant! It's required to set a bunch of CGO variables before it will compile. Setting those seems to have resolved the issue. Thanks for answering so quickly!!! Sorry the issue seemed so unrelated I didn't think to check that.

**Reply by liad-arnacio** [reply]
Hey, can you share what variables you needed to change? (For future reference)

**Reply by snowzach** [reply]
It was related to that project and not the linter. If you're still curious look at the Makefile of the project. https://github.com/estuary/flow/blob/master/Makefile They should be in there.

**Comment by ubiuser** [comment]
In my case `golangci-lint run -c .golangci.yaml` passes locally, but the GitHub action fails.
My action:
```
    - name: Run golangci-lint
      uses: golangci/golangci-lint-action@v2.5.2
      with:
        version: v1.43
        args: -c .golangci.yaml
```

**Reply by butuzov** [reply]
Do you have `cache` restore step?

**Reply by ubiuser** [reply]
I added `skip-pkg-cache` and `skip-build-cache`, but have the same result. This is from the run:

<details>

```
Run golangci/golangci-lint-action@v2.5.2
  with:
    version: v1.43
    args: -c .golangci.yaml
    github-token: ***
    skip-pkg-cache: true
    skip-build-cache: true
    only-new-issues: false
    skip-go-installation: false
```

</details>


<details>

```
go env
  GO111MODULE=""
  GOARCH="amd64"
  GOBIN=""
  GOCACHE="/home/runner/.cache/go-build"
  GOENV="/home/runner/.config/go/env"
  GOEXE=""
  GOEXPERIMENT=""
  GOFLAGS=""
  GOHOSTARCH="amd64"
  GOHOSTOS="linux"
  GOINSECURE=""
  GOMODCACHE="/home/runner/go/pkg/mod"
  GONOPROXY=""
  GONOSUMDB=""
  GOOS="linux"
  GOPATH="/home/runner/go"
  GOPRIVATE=""
  GOPROXY="https://proxy.golang.org,direct"
  GOROOT="/opt/hostedtoolcache/go/1.17.6/x64"
  GOSUMDB="sum.golang.org"
  GOTMPDIR=""
  GOTOOLDIR="/opt/hostedtoolcache/go/1.17.6/x64/pkg/tool/linux_amd64"
  GOVCS=""
  GOVERSION="go1.17.6"
  GCCGO="gccgo"
  AR="ar"
  CC="gcc"
  CXX="g++"
  CGO_ENABLED="1"
  GOMOD="/home/runner/work/payrun/payrun/go.mod"
  CGO_CFLAGS="-g -O2"
  CGO_CPPFLAGS=""
  CGO_CXXFLAGS="-g -O2"
  CGO_FFLAGS="-g -O2"
  CGO_LDFLAGS="-g -O2"
  PKG_CONFIG="pkg-config"
  GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build4078430449=/tmp/go-build -gno-record-gcc-switches"
```

</details>

<details>

```
Installed Go in 1638ms
Omitting ~/.cache/go-build from cache directories
Omitting ~/go/pkg from cache directories
Requested golangci-lint 'v1.43', using 'v1.43.0', calculation took 1669ms
Installing golangci-lint v1.43.0...
Downloading https://github.com/golangci/golangci-lint/releases/download/v1.43.0/golangci-lint-1.43.0-linux-amd64.tar.gz ...
Cache not found for input keys: golangci-lint.cache-2715-5a2cdac6f3987f07492cb70be50e74915d88a9e3, golangci-lint.cache-2715-, golangci-lint.cache-
/usr/bin/tar xz --overwrite --warning=no-unknown-keyword -C /home/runner -f /home/runner/work/_temp/f0741995-a603-4d54-b0d2-d295662f1820
Installed golangci-lint into /home/runner/golangci-lint-1.43.0-linux-amd64/golangci-lint in 395ms
Prepared env in 2066ms
```

</details>

<details>

```
run golangci-lint
  Running [/home/runner/golangci-lint-1.43.0-linux-amd64/golangci-lint run --out-format=github-actions -c .golangci.yaml] in [] ...
  Error: s.DB undefined (type *Service has no field or method DB) (typecheck)
  Error: s.DB undefined (type *Service has no field or method DB) (typecheck)
  Error: s.DB undefined (type *Service has no field or method DB) (typecheck)
  Error: s.DB undefined (type *Service has no field or method DB) (typecheck)
  Error: s.DB undefined (type *Service has no field or method DB) (typecheck)
  Error: s.DB undefined (type *Service has no field or method DB) (typecheck)
  Error: s.DB undefined (type *Service has no field or method DB) (typecheck)
  Error: s.DB undefined (type *Service has no field or method DB) (typecheck)
  Error: s.DB undefined (type *Service has no field or method DB) (typecheck)
  Error: s.DB undefined (type *Service has no field or method DB) (typecheck)
  Error: s.DB undefined (type *Service has no field or method DB) (typecheck)
  Error: s.DB undefined (type *Service has no field or method DB) (typecheck)
  Error: s.DB undefined (type *Service has no field or method DB) (typecheck)
  Error: s.DB undefined (type *Service has no field or method DB) (typecheck)
  Error: s.DB undefined (type *Service has no field or method DB) (typecheck)
  
  Error: issues found
  Ran golangci-lint in 32982ms
```

</details>


There are no any other cache steps in the workflow.

**Reply by ubiuser** [reply]
The solution was to give a name to the embedded field. So instead of
```go
type localStruct struct {
	*externalStruct
}
```
I have to write
```go
type localStruct struct {
	embedded *externalStruct
}
```
Feels like a bug to me.

**Reply by ubiuser** [reply]
The workaround above doesn't work with interfaces, so I ended up installing and manually running the linter in my workflow instead of using the official GH Action.

**Reply by butuzov** [reply]
That really looks that it depend on go version. Need to test a bit more.

**Reply by ubiuser** [reply]
@butuzov I did some more testing and I believe the issue was that the linter didn't have access to the **private repo** that implemented the structs and interfaces that were embedded in the actual project, hence it didn't find the field definitions. The solution that seemed to work is something like the `Private Repo Config` step before the linter step:
```
name: Run linters
inputs:
  token:
    description: github token
    default: ${{ github.token }}
    required: true

runs:
  using: composite

  steps:
    - name: Private Repo Config
      run: |
        git config \
          --global \
          url."https://<your-user>:${{ inputs.token }}@github.com".insteadOf \
          "https://github.com"
        go env -w GOPRIVATE=github.com/<your-private-repo>
      shell: bash

    - name: Run golangci-lint
      uses: golangci/golangci-lint-action@v2.5.2
      with:
        version: v1.43
        args: -c .golangci.yaml
        github-token: ${{ inputs.token }}
```
Does that sound like a reasonable explanation?

**Reply by hunjixin** [reply]
i have this issue,but myproject is all public,but fail too. run in local is okay, but when poster to ci, it fail

**Reply by Shugyousha** [reply]
> The solution that seemed to work is something like the Private Repo Config step before the linter step:

We already had this set up but our `golangci-lint` run was still failing with a `typecheck` error.

The source of the issue turned out to be that the API token we were using was expired (but due to the caching we only noticed the issue later). As soon as we regenerated the API token and used it in the Github Action everything started working again.

It would be good if golangci-lint could throw an error or at least a warning in case that it cannot download all the necessary dependencies.

I hope this will help someone else searching for the same issue in the future!

**Reply by Panlq** [reply]
I also encountered this problem， Running in locally is okay, but failed in Jenkins ci

**Comment by ccding** [comment]
change to `version: v1.56` can solve this issue

**Comment by zhangtianze1997** [comment]
I use golangci-lint v1.56.0 and go 1.22 in my project. The golangci-lint reports an error like `p.Mutex undefined (type *Procurator has no field or method Mutex) (typecheck)`. 
The Procurator struct is defined in pkg/procurator/procurator.go as 
```go
type Procurator struct {
	*controller.Group
	envServ *EnvServer
}
```
The Mutex field is in the anonymous field `*controller.Group` and it is defined in pkg/controller/group.go as 
```go
type Group struct {
	QemuName    string
	cfg         *common.Config
	Qoms        *qemu.QomMap
	Mutex       sync.Mutex
	Controllers map[string]*Controller
}
```

What confuses me is that the error occurs after I commit a patch in my project. The patch is independent and it does not modify anything related to these structs. But this patch imports a module in our private github server. How can I fix it?

**Reply by zhangtianze1997** [reply]
BTW, the build is successful. Only lint reports typecheck error.

**Reply by ldez** [reply]
https://golangci-lint.run/usage/faq/#why-do-you-have-typecheck-errors

**Reply by zhangtianze1997** [reply]
@ldez Thanks for your reply. I have checked the hints in the link but it cannot help me to fix the problem.

>Ensure the version of golangci-lint is built with a compatible version of Go.

The golangci-hint version is v1.56.0 which supports go 1.22.

>Ensure dependencies are up-to-date with go mod tidy.

I have run `go mod tidy` before lint.

>Ensure building works with `go run ./...`/`go build ./...` - whole package.

Yes, it is.

>Ensure you are not running an analysis on code that depends on files/packages outside the scope of the analyzed elements.

The patch imports a module in our private Github Enterprise repo. Does it cause the error? The module is imported in different package and they are independent. :(

>If using CGO, ensure all require system libraries are installed.

The build is successful with `CGO_ENABLED=1`.

**Reply by bombsimon** [reply]
Maybe dumb question but do you use `CGO_ENABLED=1` when you run the linter as well?

**Reply by zhangtianze1997** [reply]
@bombsimon Yes, I use `GO111MODULE=on CGO_ENABLED=1 /path/to/golangci-lint cache clean` and `GO111MODULE=on CGO_ENABLED=1 /path/to/golangci-lint run`

**Reply by bombsimon** [reply]
> The patch imports a module in our private Github Enterprise repo. Does it cause the error?

It should not as long as you have the dependency in your go mod directory (`GOMODCACHE`) which you must have if `go build` works. You can always run `go get all` to let the Go toolchain download your dependencies.

Given that `go build ./...` works and you have the same CGO configuration I don't have any good ideas right now without a reproducible example.

**Comment by cold-bin** [comment]
I meet the same issue

**Comment by you-fail-me** [comment]
I'm getting the same issue with code generated by Mockery getting erroneously flagged - it doesn't "see" methods inherited from `github.com/stretchr/testify/mock` which mockery's generated mocks embed. 

For some reason I'm only getting it in a docker container built for CI and not with local macOS installation. I'm using Go 1.23.4 and pre-built v1.62.2 of golangci-lint (installed via `curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b $(go env GOPATH)/bin v1.62.2`).

Could anyone advise what might be wrong here? looks like some flakiness with the linter and the errors don't really point anywhere. Couldn't infer anything helpful from the discussion above either, so would be grateful for pointers!

**Reply by bombsimon** [reply]
Are you using CGO? Do you have all C dependencies installed in your container as you do in macOS? What's the output of `go build ./...` in the container?

**Comment by sikmir** [comment]
I have the same problem, latest golangci-lint, go 1.24.

**Reply by ldez** [reply]
https://golangci-lint.run/welcome/faq/#why-do-you-have-typecheck-errors

**Comment by rigazilla** [comment]
same here:
- go 1.24.4 has the problem
- go 1.23.9 hasn't

**Reply by ldez** [reply]
https://golangci-lint.run/welcome/faq/#why-do-you-have-typecheck-errors

**Reply by rigazilla** [reply]
seems I solved updating golangci-lint to 1.64.8, previous version (1.61) was too old for golang 1.24
